src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -8657,6 +8657,51 @@ bool MergeTreeData::initializeDiskOnConfigChange(const std::set<String> & new_ad
     return true;
 }
 
+void MergeTreeData::loadPrimaryKeys()
+{
+    // Define the states of parts that need to be processed
+    DataPartStates affordable_states = { MergeTreeDataPartState::Active, MergeTreeDataPartState::Outdated, MergeTreeDataPartState::Deleting };
+
+    // Thread pool to process loading in parallel
+    auto & thread_pool = DB::getActivePartsLoadingThreadPool().get();
```

Comment:
Attempting to do it fast is ok, but I also think we'd rather make it simpler first. We can parallelize it later if it's ever needed.

For instance, as it is right now, I see non-intentional behavior. You're using the active parts loading `ThreadPool` for loading in parallel tables in `InterpreterSystemQuery::loadPrimaryKeys()`, making sure you never load more tables in parallel than the max (of threads or tables). But then, within each table, here you're using that same thread pool to make sure you never load more parts in parallel than the max (of threads or parts). Aren't you here potentially scheduling way more tasks than you initially intended?

Imagine having a `ThreadPool` with 8 threads and 100 tables where each table has 200 parts. During the first iteration in `InterpreterSystemQuery::loadPrimaryKeys()` you call `MergeTreeData::loadPrimaryKeys()` that schedules 8 parts in parallel. But then, for the next 7 iterations you do the same (assuming no previous iteration has finished yet), creating a total of 8 * (1 + 8) = 72 tasks, instead of the 8 intended. In other words, parallelizing this complicates things quite a bit.

Reply:
Thank you for the insightful feedback! I've updated the approach to remove concurrency at the table level, focusing parallelism only on loading parts within each table.

Now, the process sequentially iterates through tables and loads parts in parallel using the thread pool. This ensures we don't inadvertently create more tasks than intended by avoiding nested parallelism.

This should simplify the logic while maintaining efficient part-level concurrency where it's most needed. Let me know if this looks better!